File: frontend/src/types/pipeline.ts
export type PipelineStage =
  | "queued"
  | "sequencing"
  | "parsing_vcf"
  | "phenotype"
  | "similarity"
  | "literature"
  | "prioritization"
  | "structure"
  | "diagnosis"
  | "complete"
  | "error";

export type StageStatus = "pending" | "running" | "complete" | "error";

export interface SSEEvent {
  stage: PipelineStage;
  status: StageStatus;
  progress: number;
  message: string;
  data?: Record<string, unknown>;
}

export interface PhenotypeTerm {
  term: string;
  original: string;
  hpo_id: string;
  source: string;
  score: number;
}

export interface SimilarCase {
  disease: string;
  score: number;
  source: string;
  gene: string;
  orpha_code: string;
  omim_id: string;
  hpo_overlap: string[];
  overlap_count: number;
}

export interface LiteratureItem {
  title: string;
  authors: string;
  journal: string;
  year: string;
  pmid: string;
  doi: string;
  url: string;
  snippet: string;
  citations: number;
  source: string;
}

export interface PrioritizedVariant {
  gene: string;
  variant_id?: string;
  cdna_change: string;
  protein_change: string;
  chromosome: string;
  position: number;
  ref: string;
  alt: string;
  zygosity: string;
  gnomad_af: number;
  consequence: string;
  clinvar_significance: string;
  franklin?: string | null;
  alphamissense?: { am_pathogenicity: number; am_class: string } | null;
  is_lof: boolean;
  is_missense: boolean;
  gene_phenotype_match: boolean;
  priority_score: number;
  priority_reasons: string[];
  novel: boolean;
}

export type ACMGClassification =
  | "Pathogenic"
  | "Likely Pathogenic"
  | "Variant of Uncertain Significance"
  | "Likely Benign"
  | "Benign";

export interface ACMGCriterion {
  code: string;
  met: boolean;
  strength: string;
  description: string;
}

export interface VariantResult {
  variant_id: string;
  gene: string;
  cdna_change: string;
  protein_change: string;
  chromosome: string;
  position: number;
  ref: string;
  alt: string;
  zygosity: string;
  gnomad_af: number;
  classification: ACMGClassification;
  classification_score: number;
  criteria_met: ACMGCriterion[];
  clinical_significance: string;
  associated_diseases: string[];
  actionable: boolean;
  recommendation: string;
}

export interface DiseaseCandidate {
  rank: number;
  disease_name: string;
  orpha_code: string;
  omim_id?: string;
  score: number;
  phenotype_match_score: number;
  genotype_match_score: number;
  prevalence: string;
  inheritance_pattern: string;
  matched_symptoms: string[];
  unmatched_symptoms: string[];
  supporting_genes: string[];
  reasoning: string;
}

export interface DeepRareResult {
  candidates: DiseaseCandidate[];
  total_variants_analyzed: number;
  phenotype_terms_matched: number;
  model_version: string;
  confidence_note: string;
}

export interface ACMGResult {
  variants: VariantResult[];
  pathogenic_count: number;
  likely_pathogenic_count: number;
  vus_count: number;
  benign_count: number;
  actionable_variants: string[];
  classifier_version: string;
}

export interface StructuralImpact {
  impact_type: string;
  severity: "High" | "Medium" | "Low";
  affected_domain: string;
  description: string;
}

export interface ProteinStructure {
  gene: string;
  uniprot_id: string;
  pdb_id?: string;
  structure_url: string;
  plddt_score: number;
  variant_position: number;
  wild_type_residue: string;
  mutant_residue: string;
}

export interface AlphaFoldResult {
  gene: string;
  variant: string;
  wild_type_structure: ProteinStructure;
  mutant_structure: ProteinStructure;
  rmsd: number;
  structural_impacts: StructuralImpact[];
  pathogenicity_upgrade: boolean;
  upgraded_from?: string;
  upgraded_to?: string;
  functional_summary: string;
  pdb_wild_type: string;
  pdb_mutant: string;
}

export interface PipelineResult {
  session_id: string;
  patient_name: string;
  deeprare: DeepRareResult;
  acmg: ACMGResult;
  alphafold: AlphaFoldResult[];
  summary: string;
  time_to_diagnosis_estimate: string;
  report_url?: string;
}

File: backend/app/models/pipeline.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from enum import Enum


class PipelineStage(str, Enum):
    QUEUED = "queued"
    PARSING_VCF = "parsing_vcf"
    DEEPRARE = "deeprare"
    ACMG = "acmg"
    ALPHAFOLD = "alphafold"
    GENERATING_REPORT = "generating_report"
    COMPLETE = "complete"
    ERROR = "error"


class StageStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETE = "complete"
    ERROR = "error"


class SSEEvent(BaseModel):
    stage: PipelineStage
    status: StageStatus
    progress: int  # 0-100
    message: str
    data: Optional[Dict[str, Any]] = None


# ─── DeepRare Results ─────────────────────────────────────────────────────────

class DiseaseCandidate(BaseModel):
    rank: int
    disease_name: str
    orpha_code: str
    omim_id: Optional[str] = None
    score: float                    # 0.0 – 1.0
    phenotype_match_score: float
    genotype_match_score: float
    prevalence: str
    inheritance_pattern: str
    matched_symptoms: List[str]
    unmatched_symptoms: List[str]
    supporting_genes: List[str]
    reasoning: str


class DeepRareResult(BaseModel):
    candidates: List[DiseaseCandidate]
    total_variants_analyzed: int
    phenotype_terms_matched: int
    model_version: str = "DeepRare-v2.1-mock"
    confidence_note: str


# ─── ACMG Results ─────────────────────────────────────────────────────────────

class ACMGClassification(str, Enum):
    PATHOGENIC = "Pathogenic"
    LIKELY_PATHOGENIC = "Likely Pathogenic"
    VUS = "Variant of Uncertain Significance"
    LIKELY_BENIGN = "Likely Benign"
    BENIGN = "Benign"


class ACMGCriterion(BaseModel):
    code: str          # e.g. PVS1, PS1, PM2
    met: bool
    strength: str      # Pathogenic_VeryStrong / Pathogenic_Strong / etc.
    description: str


class VariantResult(BaseModel):
    variant_id: str
    gene: str
    cdna_change: str
    protein_change: str
    chromosome: str
    position: int
    ref: str
    alt: str
    zygosity: str
    gnomad_af: float          # allele frequency
    classification: ACMGClassification
    classification_score: int  # net score
    criteria_met: List[ACMGCriterion]
    clinical_significance: str
    associated_diseases: List[str]
    actionable: bool
    recommendation: str


class ACMGResult(BaseModel):
    variants: List[VariantResult]
    pathogenic_count: int
    likely_pathogenic_count: int
    vus_count: int
    benign_count: int
    actionable_variants: List[str]
    classifier_version: str = "ACMG-2015-mock"


# ─── AlphaFold3 Results ───────────────────────────────────────────────────────

class ProteinStructure(BaseModel):
    gene: str
    uniprot_id: str
    pdb_id: Optional[str] = None
    structure_url: str             # URL or base64 PDB data
    plddt_score: float             # confidence 0-100
    variant_position: int
    wild_type_residue: str
    mutant_residue: str


class StructuralImpact(BaseModel):
    impact_type: str               # e.g. "Domain disruption", "Disulfide bond loss"
    severity: str                  # High / Medium / Low
    affected_domain: str
    description: str


class AlphaFoldResult(BaseModel):
    gene: str
    variant: str
    wild_type_structure: ProteinStructure
    mutant_structure: ProteinStructure
    rmsd: float                    # Å deviation between WT and mutant
    structural_impacts: List[StructuralImpact]
    pathogenicity_upgrade: bool    # did this upgrade a VUS?
    upgraded_from: Optional[str] = None
    upgraded_to: Optional[str] = None
    functional_summary: str
    pdb_wild_type: str             # inline PDB data for viewer
    pdb_mutant: str


# ─── Full Pipeline Result ─────────────────────────────────────────────────────

class PipelineResult(BaseModel):
    session_id: str
    patient_name: str
    deeprare: DeepRareResult
    acmg: ACMGResult
    alphafold: List[AlphaFoldResult]
    summary: str
    time_to_diagnosis_estimate: str
    report_url: Optional[str] = None
